crates/turbopack-ecmascript/src/analyzer/mod.rs

```diff
@@ -2852,6 +2873,97 @@ impl WellKnownObjectKind {
     }
 }
 
+#[derive(Debug, Clone)]
+pub struct RequireContextOptions {
+    pub dir: String,
+    pub include_subdirs: bool,
+    /// this is a regex (pattern, flags)
+    pub filter: Regex,
+}
+
+pub fn regex_from_js(pattern: &str, flags: &str) -> Result<Regex, &'static str> {
+    // rust regex doesn't allow escaped slashes, but they are necessary in js
+    let pattern = pattern.replace("\\/", "/");
+
+    let mut applied_flags = String::new();
+    for flag in flags.chars() {
+        match flag {
+            // case-insensitive: letters match both upper and lower case
+            'i' => applied_flags.push('i'),
+            // multi-line mode: ^ and $ match begin/end of line
+            'm' => applied_flags.push('m'),
+            // allow . to match \n
+            's' => applied_flags.push('s'),
+            // Unicode support (enabled by default)
+            'u' => applied_flags.push('u'),
+            // default in rust, ignore
+            'g' => {}
+            // TODO: maybe error in the future
+            _ => {}
+        }
+    }
+
+    let regex = if !applied_flags.is_empty() {
+        format!("(?{}){}", applied_flags, pattern)
+    } else {
+        pattern
+    };
+
+    Regex::new(&regex).map_err(|_| "could not convert javascript regex to rust regex")
+}
+
+pub fn parse_require_context(args: &Vec<JsValue>) -> Result<RequireContextOptions, &'static str> {
+    if !(1..=3).contains(&args.len()) {
+        return Err("only 1-3 arguments are supported (mode is not supported)");
+    }
+
+    let Some(dir) = args[0].as_str().map(|s| s.to_string()) else {
+        return Err("dir needs to be a constant argument");
+    };
+
+    let include_subdirs = if let Some(include_subdirs) = args.get(1) {
+        if let Some(include_subdirs) = include_subdirs.as_bool() {
+            include_subdirs
+        } else {
+            return Err("includeSubdirs needs to be a boolean");
+        }
+    } else {
+        true
+    };
+
+    let filter = if let Some(filter) = args.get(2) {
+        if let JsValue::Constant(ConstantValue::Regex(pattern, flags)) = filter {
+            regex_from_js(pattern, flags)?
+        } else {
+            return Err("filter needs to be a regex");
+        }
+    } else {
+        // https://webpack.js.org/api/module-methods/#requirecontext
+        // > optional, default /^\.\/.*$/, any file
+        Regex::new("^\\./.*$").expect("valid static regex should always compile")
+    };
+
+    Ok(RequireContextOptions {
+        dir,
+        include_subdirs,
+        filter,
+    })
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct RequireContextValue {
+    pub(crate) map: IndexMap<String, String>,
+}
+
+impl Hash for RequireContextValue {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        for (k, v) in self.map.iter() {
```

Comment:
Should hash something even in the case it's empty (e.g. the length) so consecutive values don't hash the same [{}, { k: 1 }], [{ k: 1 }, {}].

Reply:
hmm, I can add the length, but I don't think it would prevent your example from hashing differently